syntax = "proto2";

package foundation;

import "geometry.proto";

message EntityCollection {
  required string id = 1;
  repeated string entity_ids = 2;
}

message Entity {
  required string id = 1;
  optional EntityCollection children_id = 2;
  oneof type_specific {
    Word word = 3;
    FillerString filler_string = 4;
    SubWord sub_word = 5;
    Page page = 6;
    Text text = 7;
  }
}

message Word {
  required BBox bbox = 1;
  required string text = 2;
}

message FillerString { required string text = 1; }

message SubWord {
  required string word_id = 1;
  required uint32 start_index = 2;
  required uint32 end_index = 3;
}

message Page {
  repeated BBox bbox = 1;
  required string image_path = 2;
}

message Text {
  optional EntityCollection words_id = 1;
  optional double likeness_score = 2;
  oneof type_specific {
    Date date = 3;
    CurrencyAmount currency_amount = 4;
    PersonName person_name = 5;
    Address address = 6;
  }
}

message Date {
  required uint32 year = 1;
  required uint32 month = 2;
  required uint32 day = 3;
}

message CurrencyAmount {
  // ISO 4217 Currency codes
  enum Currency { USD = 840; }
  required Currency currency = 1;
  required int64 amount = 2; // amounts are always in ISO 4217 minor amounts
}

message PersonName {}

message Address {}